Render markdown in entity descriptions

xsebek opened the pull request:
* use `Markdown.Document` as `entityDescription`
* add missing spaces in `chunksOf`
* fix code in `entities.yaml` (mostly types and few outdated snippets)
* add code markdown in craft tutorial
* use colours for types and entities

- closes #1408
- closes #1409

xsebek commented on data/entities.yaml:
```diff
@@ -1323,16 +1322,17 @@
+    to compute a final result of type `b`{=type}.  For example, if
+    `x`{=snippet}, `y`{=snippet}, and `z`{=snippet}
+    are nearby actors, then `meetAll f b0`{=snippet} is equivalent to
+    `b1 <- f b0 x; b2 <- f b1 y; f b2 z`{=snippet}.
```
It would be kind of neat if we could introduce variables with markup and then apply them to the rest of the document. Something like:
```markdown
If `x`{=variable} is a nearby actor, then `give "apple" x` will end all hunger.
```

byorgey replied:
Yes, that would be neat, and not too hard in theory, though it might require a bit of work to plumb the right information around.  The basic idea would be that when we typecheck code in backticks we don't simply start it in an empty context, but rather start it in a context that already contains any declared variables (with types like `forall a. a`).

kostmo replied:
[Literate Swarm](https://wiki.haskell.org/Literate_programming)?

xsebek commented:
This is how I debugged it:
```Haskell
ghci> -- parse entities.yaml
ghci> em :: [Entity] <- either (fail . prettyPrintParseException) pure =<< decodeFileEither "data/entities.yaml"
ghci> -- format 3D printer descriptions like in TUI
ghci> let Just d3 = L.find (\e -> view entityName e == T.pack "3D printer") em in chunksOf 99 . toStream $ view entityDescription d3
```
And it finally seems to look OK:
<img width="800" alt="Screenshot 2023-08-06 at 2 19 12" src="https://github.com/swarm-game/swarm/assets/44544735/6d40a21c-9c2b-493b-bfd7-58f1135f5621">

kostmo commented on data/entities.yaml:
```diff
@@ -1374,9 +1374,9 @@
+    is a value of type `[x : int, y : text]`{=type}.  Individual fields
     can be projected using dot notation.  For example,
-    `let r = [y="hi", x=2] in r.x` has the value 2.  The order
+    `let r = [y="hi", x=2] in r.x` has the value `2`.  The order
```
Annotate snippet?
```suggestion
    `let r = [y="hi", x=2] in r.x`{=snippet} has the value `2`.  The order
```

xsebek replied:
Not if we want it to be parsed, maybe I should have named it `{=invalid}` or `{=non-swarm}`.

kostmo commented on src/Swarm/Language/Text/Markdown.hs:
```diff
@@ -181,7 +216,8 @@ fromTextPure :: Text -> Either String (Document Text)
   let runSimple = left show . runIdentity
-  runSimple $ Mark.commonmarkWith spec "markdown" t
+  Document tokenizedDoc <- runSimple $ Mark.commonmarkWith spec "markdown" t
+  return . Document $ normalise <$> tokenizedDoc
```
Random thought: would there be any value in having a `NormalizedDocument` newtype whose "smart constructor" performs the `normalize` function?

xsebek replied:
Honestly, if `Document` used `Seq` I would consider doing it immediately in `(<>)`, the naive approach here merges the text one-by-one anyway.

byorgey commented:
> And it finally seems to look OK:

The formatting looks great, but now it looks cramped.  Can we add blank lines in between every paragraph, like we used to?

byorgey commented on data/entities.yaml:
```diff
@@ -797,7 +795,7 @@
     char: '%'
   description:
-  - A "tape drive" allows you to `backup`; that is, to `drive` in reverse.
+  - A "tape drive" allows you to `backup`; that is, to drive in reverse.
```
```suggestion
  - A `tape drive`{=entity} allows you to `backup`; that is, to drive in reverse.
```

byorgey commented:
CI is failing since there is some code in a docspec test in `Swarm.Game.Exception` that needs to be fixed, I think.

xsebek commented:
> The formatting looks great, but now it looks cramped. Can we add blank lines in between every paragraph, like we used to?

@byorgey should be fixed now, I had to rethink the code a little and keep paragraphs for printing.

xsebek commented:
Btw. the docspec test would not fail if we had `OverloadedLists` extension enabled, but as it is I used `mempty`.

But for markdown in Haskell code (e.g. generated entity/robot description) we can use that instance.

kostmo commented on src/Swarm/Language/Text/Markdown.hs:
```diff
@@ -208,11 +251,8 @@ unStream = \case
   TextNode a t -> (TextNode a, t)
   CodeNode t -> (CodeNode, t)
   RawNode a t -> (RawNode a, t)
-  ParagraphBreak -> error "Logic error: Paragraph break can not be unstreamed!"
```
Always nice to get rid of partial functions...

xsebek replied:
Indeed, I don't think that paragraph break worked anyway! 😄 It shows that I wrote it before knowing where and how we use the documents.

kostmo commented on src/Swarm/Language/Text/Markdown.hs:
```diff
@@ -230,13 +269,16 @@ chunksOf n = chop (splitter True n)
               -- In case single word (e.g. web link) does not fit on line we must put
               -- it there and guarantee progress (otherwise chop will cycle)
-              if start then ([ww], wws) else ([], ws)
+              if start then ([T.take i ww], T.drop i ww : wws) else ([], ws)
```
Can we utilize `T.splitAt i ww` here?

xsebek replied:
Maybe, but I doubt it would make the code simpler or more performant.

kostmo replied:
Sure, it's up to your preference.  I suppose it could look like this.
```suggestion
              if start then bimap pure (: wws) $ T.splitAt i ww else ([], ws)
```

xsebek replied:
Thanks, I was about to try write it myself, but this is definitely harder to read for me. 😆

kostmo commented on src/Swarm/Language/Text/Markdown.hs:
````diff
@@ -265,15 +306,36 @@ streamToText = T.concat . map nodeToText
+  LeafCode c -> wrap "`" (prettyText c)
+  LeafCodeBlock f c -> codeBlock f $ prettyText c
+ where
+  codeBlock f t = "```" <> T.pack f <> "\n" <> t <> "\n```"
````
```suggestion
  codeBlock f t = wrap "```" $ T.pack f <> "\n" <> t <> "\n"
```

kostmo reviewed (approved):
LGTM with minor suggestions.  Also don't forget [this one](https://github.com/swarm-game/swarm/pull/1413/files#r1285142879) from @byorgey.

xsebek commented:
@kostmo I went ahead and added some colours anyway:
<img width="800" alt="Screenshot 2023-08-12 at 12 04 07" src="https://github.com/swarm-game/swarm/assets/44544735/14f84d9c-e9b8-470b-901f-0d657c637919">

I also added a script to run scenarios _in order_ with `--autoplay` and show goal descriptions. 😁 
```
❯ scripts/autoplay-tutorials.sh
backstory.yaml	CONTINUE [Y/n]:
move.yaml	CONTINUE [Y/n]:
craft.yaml	CONTINUE [Y/n]:
grab.yaml	CONTINUE [Y/n]:
place.yaml	CONTINUE [Y/n]: n
```
